File: VB/NavigateCommandExample/ViewModel/MainViewModel.vb
#Region "#usings"
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Xpf.Scheduling
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Linq
#End Region ' #usings

Namespace SchedulerDragDropExample
    <POCOViewModel> _
    Public Class MainViewModel
        Private Const DayViewVisibleDaysCount As Integer = 7
        Public Overridable Property Appointments() As ObservableCollection(Of MedicalAppointment)
        Public Overridable Property Doctors() As ObservableCollection(Of Doctor)
        Public Overridable Property PaymentStates() As ObservableCollection(Of PaymentState)
        Public Overridable Property SchedulerStart() As Date

        Protected Sub New()
            Doctors = New ObservableCollection(Of Doctor)()
            Appointments = New ObservableCollection(Of MedicalAppointment)()
            PaymentStates = New ObservableCollection(Of PaymentState)()

            SchedulerStart = Date.Today.Date

            CreateDoctors()
            CreateAppointments()
            CreatePaymentStates()
        End Sub

        Private Sub CreateAppointments()
            Dim rand As New Random(Date.Now.Millisecond)
            Appointments.Add(MedicalAppointment.Create(startTime:= Date.Now.Date.AddHours(10), endTime:= Date.Now.Date.AddHours(11.5), doctorId:= 1, paymentStateId:= 1, location:= "101", patientName:= "Dave Murrel", note:= "Take care", firstVisit:= True))
            Appointments.Add(MedicalAppointment.Create(startTime:= Date.Now.Date.AddDays(2).AddHours(15), endTime:= Date.Now.Date.AddDays(2).AddHours(16.5), doctorId:= 1, paymentStateId:= 1, location:= "101", patientName:= "Mike Roller", note:= "Schedule next visit soon", firstVisit:= True))

            Appointments.Add(MedicalAppointment.Create(startTime:= Date.Now.Date.AddDays(1).AddHours(11), endTime:= Date.Now.Date.AddDays(1).AddHours(12), doctorId:= 2, paymentStateId:= 1, location:= "103", patientName:= "Bert Parkins", note:= String.Empty, firstVisit:= True))
            Appointments.Add(MedicalAppointment.Create(startTime:= Date.Now.Date.AddDays(2).AddHours(10), endTime:= Date.Now.Date.AddDays(2).AddHours(12), doctorId:= 2, paymentStateId:= 0, location:= "103", patientName:= "Carl Lucas", note:= String.Empty, firstVisit:= False))

            Appointments.Add(MedicalAppointment.Create(startTime:= Date.Now.Date.AddHours(12), endTime:= Date.Now.Date.AddHours(13.5), doctorId:= 3, paymentStateId:= 1, location:= "104", patientName:= "Brad Barnes", note:= "Tests are necessary", firstVisit:= False))
            Appointments.Add(MedicalAppointment.Create(startTime:= Date.Now.Date.AddDays(1).AddHours(14), endTime:= Date.Now.Date.AddDays(1).AddHours(15), doctorId:= 3, paymentStateId:= 1, location:= "104", patientName:= "Richard Fisher", note:= String.Empty, firstVisit:= True))
        End Sub

        Private Sub CreateDoctors()
            Doctors.Add(Doctor.Create(id:= 1, name:= "Stomatologist"))
            Doctors.Add(Doctor.Create(id:= 2, name:= "Ophthalmologist"))
            Doctors.Add(Doctor.Create(id:= 3, name:= "Surgeon"))
        End Sub

        Private Sub CreatePaymentStates()
            PaymentStates.Add(PaymentState.Create(id:= 0, caption:= "Unpaid", color:= "Tomato"))
            PaymentStates.Add(PaymentState.Create(id:= 1, caption:= "Paid", color:= "LightGreen"))
        End Sub

        #Region "#MyMove"
        Public Sub MyMoveLeft(ByVal scheduler As SchedulerControl)
            MyMove(False, scheduler)
        End Sub
        Public Sub MyMoveRight(ByVal scheduler As SchedulerControl)
            MyMove(True, scheduler)
        End Sub
        Public Sub MyMove(ByVal forward As Boolean, ByVal scheduler As SchedulerControl)
            If SchedulerStart.Date = scheduler.Start.Date Then
                Return
            End If
            FixVisibleDays(forward, scheduler)
            SchedulerStart = scheduler.Start
        End Sub
        Private Sub FixVisibleDays(ByVal forward As Boolean, ByVal scheduler As SchedulerControl)
            Dim myView As DayView = TryCast(scheduler.ActiveView, DayView)
            Dim startDate As Date = scheduler.Start
            If myView IsNot Nothing Then
                Dim selection As DateTimeRange = scheduler.SelectedInterval
                startDate = CorrectDate(startDate, forward)
                Dim visibleDays As List(Of Date) = GetDays(startDate)
                ' Specify DayView's visible days 
                myView.Days = visibleDays
                scheduler.SelectedInterval = AdjustSelection(forward, selection)
            End If
        End Sub
        #End Region ' #MyMove

        Public Sub MyGoToDate(ByVal scheduler As SchedulerControl)
            scheduler.ShowGotoDateWindow(Date.Today)
            FixVisibleDays(True, scheduler)
            SchedulerStart = scheduler.Start
        End Sub

        Public Sub MyGoToToday(ByVal scheduler As SchedulerControl)
            Dim startDate As Date = Date.Today
            scheduler.Start = startDate
            FixVisibleDays(True, scheduler)
            SchedulerStart = scheduler.Start
        End Sub
        #Region "#OnVisibleIntervalsChanged"
        Public Sub OnVisibleIntervalsChanged(ByVal scheduler As SchedulerControl, ByVal args As VisibleIntervalsChangedEventArgs)
            If Not NeedCorrection(args.VisibleDates.ToList()) Then
                Return
            End If
            FixVisibleDays(True, scheduler)
        End Sub
        #End Region ' #OnVisibleIntervalsChanged
        Private Function AdjustSelection(ByVal forward As Boolean, ByVal selection As DateTimeRange) As DateTimeRange
            Dim dtStart As Date
            Dim dtEnd As Date

            Dim diff As TimeSpan = selection.End - selection.Start
            dtStart = CorrectDate(selection.Start, forward)
            dtEnd = dtStart.Add(diff)
            Return New DateTimeRange(dtStart, dtEnd)
        End Function

        Public Sub MyNavigateBackward(ByVal scheduler As SchedulerControl)
            Dim myView As DayView = TryCast(scheduler.ActiveView, DayView)
            If myView IsNot Nothing Then
                scheduler.Start = scheduler.Start.AddDays(-DayViewVisibleDaysCount)
                myView.Days = GetDays(scheduler.Start)
            Else
                scheduler.ActiveView.NavigateBackward()
            End If
        End Sub
        #Region "#MyNavigateForward"
        Public Sub MyNavigateForward(ByVal scheduler As SchedulerControl)
            Dim myView As DayView = TryCast(scheduler.ActiveView, DayView)
            If myView IsNot Nothing Then
                scheduler.Start = scheduler.Start.AddDays(DayViewVisibleDaysCount)
                ' Calculate a list of days to display, excluding non-working days.
                myView.Days = GetDays(scheduler.Start)
            Else
                scheduler.ActiveView.NavigateForward()
            End If
        End Sub
        #End Region ' #MyNavigateForward
        Private Function CorrectDate(ByVal day As Date, ByVal forward As Boolean) As Date
            Dim direction As Integer
            Select Case day.DayOfWeek
                Case DayOfWeek.Saturday
                    direction = If(forward, 2, -1)
                    day = day.AddDays((direction))
                Case DayOfWeek.Sunday
                    direction = If(forward, 1, -2)
                    day = day.AddDays(direction)
            End Select
            Return day
        End Function

        Private Function NeedCorrection(ByVal days As List(Of Date)) As Boolean
            If days.Any(Function(day) (day.DayOfWeek = DayOfWeek.Saturday OrElse day.DayOfWeek = DayOfWeek.Sunday)) Then
                Return True
            End If
            Return False
        End Function
        Public Function GetDays(ByVal startDate As Date) As List(Of Date)
            Dim list = New List(Of Date)()
            Dim i As Integer = 0
            Do While list.Count < DayViewVisibleDaysCount
                Dim day = startDate.AddDays(i)
                If day.DayOfWeek <> DayOfWeek.Saturday AndAlso day.DayOfWeek <> DayOfWeek.Sunday Then
                    list.Add(day)
                End If
                i += 1
            Loop
            Return list
        End Function
    End Class
End Namespace

File: CS/NavigateCommandExample/ViewModel/MainViewModel.cs
using DevExpress.Mvvm;
using DevExpress.Xpf.Scheduling;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace SchedulerDragDropExample {
    public class MainViewModel : ViewModelBase {
        const int DayViewVisibleDaysCount = 7;
        public ObservableCollection<MedicalAppointment> Appointments { get; set; }
        public ObservableCollection<Doctor> Doctors { get; set; }
        public ObservableCollection<PaymentState> PaymentStates { get; set; }
        public DateTime SchedulerStart { get; set; }

        public MainViewModel() {
            Doctors = new ObservableCollection<Doctor>();
            Appointments = new ObservableCollection<MedicalAppointment>();
            PaymentStates = new ObservableCollection<PaymentState>();

            SchedulerStart = DateTime.Today.Date;

            CreateDoctors();
            CreateAppointments();
            CreatePaymentStates();
        }

        private void CreateAppointments() {
            Random rand = new Random(DateTime.Now.Millisecond);
            Appointments.Add(new MedicalAppointment(startTime: DateTime.Now.Date.AddHours(10), endTime: DateTime.Now.Date.AddHours(11.5), doctorId: 1, paymentStateId: 1, location: "101", patientName: "Dave Murrel", note: "Take care", firstVisit: true));
            Appointments.Add(new MedicalAppointment(startTime: DateTime.Now.Date.AddDays(2).AddHours(15), endTime: DateTime.Now.Date.AddDays(2).AddHours(16.5), doctorId: 1, paymentStateId: 1, location: "101", patientName: "Mike Roller", note: "Schedule next visit soon", firstVisit: true));

            Appointments.Add(new MedicalAppointment(startTime: DateTime.Now.Date.AddDays(1).AddHours(11), endTime: DateTime.Now.Date.AddDays(1).AddHours(12), doctorId: 2, paymentStateId: 1, location: "103", patientName: "Bert Parkins", note: string.Empty, firstVisit: true));
            Appointments.Add(new MedicalAppointment(startTime: DateTime.Now.Date.AddDays(2).AddHours(10), endTime: DateTime.Now.Date.AddDays(2).AddHours(12), doctorId: 2, paymentStateId: 0, location: "103", patientName: "Carl Lucas", note: string.Empty, firstVisit: false));

            Appointments.Add(new MedicalAppointment(startTime: DateTime.Now.Date.AddHours(12), endTime: DateTime.Now.Date.AddHours(13.5), doctorId: 3, paymentStateId: 1, location: "104", patientName: "Brad Barnes", note: "Tests are necessary", firstVisit: false));
            Appointments.Add(new MedicalAppointment(startTime: DateTime.Now.Date.AddDays(1).AddHours(14), endTime: DateTime.Now.Date.AddDays(1).AddHours(15), doctorId: 3, paymentStateId: 1, location: "104", patientName: "Richard Fisher", note: string.Empty, firstVisit: true));
        }

        private void CreateDoctors() {
            Doctors.Add(new Doctor(id: 1, name: "Stomatologist"));
            Doctors.Add(new Doctor(id: 2, name: "Ophthalmologist"));
            Doctors.Add(new Doctor(id: 3, name: "Surgeon"));
        }

        private void CreatePaymentStates() {
            PaymentStates.Add(new PaymentState(id: 0, caption: "Unpaid", color: "Tomato"));
            PaymentStates.Add(new PaymentState(id: 1, caption: "Paid", color: "LightGreen"));
        }

        #region #MyMove
        public void MyMoveLeft(SchedulerControl scheduler) {
            MyMove(false, scheduler);
        }
        public void MyMoveRight(SchedulerControl scheduler) {
            MyMove(true, scheduler);
        }
        public void MyMove(bool forward, SchedulerControl scheduler) {
            if (SchedulerStart.Date == scheduler.Start.Date) return;
            FixVisibleDays(forward, scheduler);
            SchedulerStart = scheduler.Start;
        }
        private void FixVisibleDays(bool forward, SchedulerControl scheduler) {
            DayView myView = scheduler.ActiveView as DayView;
            DateTime startDate = scheduler.Start;
            if (myView != null) {
                DateTimeRange selection = scheduler.SelectedInterval;
                startDate = CorrectDate(startDate, forward);
                List<DateTime> visibleDays = GetDays(startDate);
                // Specify DayView's visible days 
                myView.Days = visibleDays;
                scheduler.SelectedInterval = AdjustSelection(forward, selection);
            }
        }
        #endregion #MyMove

        public void MyGoToDate(SchedulerControl scheduler) {
            scheduler.ShowGotoDateWindow(DateTime.Today);
            FixVisibleDays(true, scheduler);
            SchedulerStart = scheduler.Start;
        }

        public void MyGoToToday(SchedulerControl scheduler) {
            DateTime startDate = DateTime.Today;
            scheduler.Start = startDate;
            FixVisibleDays(true, scheduler);
            SchedulerStart = scheduler.Start;
        }
        #region #OnVisibleIntervalsChanged
        public void OnVisibleIntervalsChanged(SchedulerControl scheduler, VisibleIntervalsChangedEventArgs args) {
            if (!NeedCorrection(args.VisibleDates.ToList())) return;
            FixVisibleDays(true, scheduler);
        }
        #endregion #OnVisibleIntervalsChanged
        private DateTimeRange AdjustSelection(bool forward, DateTimeRange selection) {
            DateTime dtStart;
            DateTime dtEnd;

            TimeSpan diff = selection.End - selection.Start;
            dtStart = CorrectDate(selection.Start, forward);
            dtEnd = dtStart + diff;
            return new DateTimeRange(dtStart, dtEnd);
        }

        public void MyNavigateBackward(SchedulerControl scheduler) {
            DayView myView = scheduler.ActiveView as DayView;
            if (myView != null) {
                scheduler.Start = scheduler.Start.AddDays(-DayViewVisibleDaysCount);
                myView.Days = GetDays(scheduler.Start);
            } else
                scheduler.ActiveView.NavigateBackward();
        }
        #region #MyNavigateForward
        public void MyNavigateForward(SchedulerControl scheduler) {
            DayView myView = scheduler.ActiveView as DayView;
            if (myView != null) {
                scheduler.Start = scheduler.Start.AddDays(DayViewVisibleDaysCount);
                // Calculate a list of days to display, excluding non-working days.
                myView.Days = GetDays(scheduler.Start);
            } else
                scheduler.ActiveView.NavigateForward();
        }
        #endregion #MyNavigateForward
        private DateTime CorrectDate(DateTime day, bool forward) {
            int direction;
            switch (day.DayOfWeek) {
                case DayOfWeek.Saturday:
                    direction = forward ? 2 : -1;
                    day = day.AddDays((direction));
                break;
                case DayOfWeek.Sunday:
                    direction = forward ? 1 : -2;
                    day = day.AddDays(direction);
                break;
            }
            return day;
        }

        private bool NeedCorrection(List<DateTime> days) {
            if (days.Any(day => (day.DayOfWeek == DayOfWeek.Saturday || day.DayOfWeek == DayOfWeek.Sunday)))
                return true;
            return false;
        }
        public List<DateTime> GetDays(DateTime startDate) {
            var list = new List<DateTime>();
            int i = 0;
            while (list.Count < DayViewVisibleDaysCount) {
                var day = startDate.AddDays(i);
                if (day.DayOfWeek != DayOfWeek.Saturday && day.DayOfWeek != DayOfWeek.Sunday)
                    list.Add(day);
                i++;
            }
            return list;
        }
        public void InitDays(SchedulerControl scheduler) {
            DayView myView = scheduler.ActiveView as DayView;
            if (myView != null) {
                myView.Days = GetDays(scheduler.Start);
            }
        }
    }
}
